ion-schema/src/model/constraints/timestamp_precision.rs

```diff
@@ -0,0 +1,296 @@
+    const NANOS: &'static str = "nanosecond";
+
+    /// Returns the [TimestampPrecisionValue] corresponding to a given [i64] value.
+    ///
+    /// This is for internal use only, and will panic if used incorrectly.
+    fn from_i64(value: i64) -> Self {
+        match value {
+            -4 => Year,
+            -3 => Month,
+            -2 => Day,
+            -1 => Minute,
+            0 => Second,
+            3 => Millisecond,
+            6 => Microsecond,
+            9 => Nanosecond,
+            i => unreachable!("reaching this branch indicates a programming error; value was {i}"),
+        }
+    }
+
+    fn to_i64(self) -> i64 {
+        self as i64
+    }
+
+    fn as_str(&self) -> &'static str {
+        match self {
+            Year => TimestampPrecisionValue::YEAR,
+            Month => TimestampPrecisionValue::MONTH,
+            Day => TimestampPrecisionValue::DAY,
+            Minute => TimestampPrecisionValue::MINUTE,
+            Second => TimestampPrecisionValue::SECOND,
+            Millisecond => TimestampPrecisionValue::MILLIS,
+            Microsecond => TimestampPrecisionValue::MICROS,
+            Nanosecond => TimestampPrecisionValue::NANOS,
+        }
+    }
+}
+
+impl<V: IslVersion> ReadFromIsl<V> for TimestampPrecisionValue {
+    fn try_read(ion: &Element, ctx: &LoaderContext<V>) -> IonSchemaResult<Self> {
+        match ion.as_symbol_text() {
+            Some(TimestampPrecisionValue::YEAR) => Ok(Year),
+            Some(TimestampPrecisionValue::MONTH) => Ok(Month),
+            Some(TimestampPrecisionValue::DAY) => Ok(Day),
+            Some(TimestampPrecisionValue::MINUTE) => Ok(Minute),
+            Some(TimestampPrecisionValue::SECOND) => Ok(Second),
+            Some(TimestampPrecisionValue::MILLIS) => Ok(Millisecond),
+            Some(TimestampPrecisionValue::MICROS) => Ok(Microsecond),
+            Some(TimestampPrecisionValue::NANOS) => Ok(Nanosecond),
+            _ => invalid_schema_error(format!("not a valid timestamp precision value: {ion}")),
+        }
+    }
+}
+
+impl Display for TimestampPrecisionValue {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl WriteAsIon for TimestampPrecisionValue {
+    fn write_as_ion<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
+        writer.write_symbol(self.as_str())
+    }
+}
+
+/// Represents the `timestamp_precision` constraint (ref. [ISL 1.0], [ISL 2.0]).
+///
+/// [ISL 1.0]: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#timestamp_precision
+/// [ISL 2.0]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#timestamp_precision
+#[derive(Debug, PartialEq, Clone)]
+pub struct TimestampPrecision {
+    range: IonSchemaRange<i64>,
+}
+
+impl TimestampPrecision {
+    pub(crate) fn new<T: Into<IonSchemaRange<TimestampPrecisionValue>>>(range: T) -> Self {
+        let tsp_range: IonSchemaRange<TimestampPrecisionValue> = range.into();
```

Comment:
I think I am msising something here. Is the current implementation only supporting valid ranges being constructed that are passed here. Are we only allowing ranges with same start and end. (i..e range:[year, year]).

Reply:
> Are we only allowing ranges with same start and end

No, there are are `From<_>` implementations for `IonSchemaRange` [here](https://github.com/amazon-ion/ion-schema-rust/pull/235/files#diff-b1b4a4530b9151e7a5be3879ff8c98f5ae3f7701fd84d21f0f4a1345b399b2eaR216-R231) that make it possible to do something like `TimestampPrecision::new(Year..=Day)` (equivalent to `range::[year, day]`) or `builder.timestamp_precision(Second..)` (equivalent to `range::[second, max]`).